Updates to Python bindings: std::vector, stubs

ManifoldFR opened the pull request:
In eigenpy 2.8.0 we merged a new feature from PR stack-of-tasks/eigenpy#325 porting the new version of `StdVectorPythonVisitor` from Pinocchio's upcoming version `3.x` to eigenpy itself. This PR updates the code in `std-containers.cpp` to use this code when this new eigenpy is available. Otherwise: it falls back to using the header from Pinocchio (version 3 or 2).

This PR also:
* updates the pre-commit config to *not* sort includes anymore
* uses `eigenpy::enableEigenPySpecific<Eigen::VectorXi>()` instead of the macro `ENABLE_SPECIFIC_MATRIX_TYPE` (which is not namespaced and doesn't say it comes from eigenpy)
* fixes some warning about structs being declared as `class`
* fixes a possible double-export issue coming from `walk-with-traj`
* add Python stubs using pybind11-stubgen (from recent cmake versions)
* removes includes of boost/python in favour of using eigenpy/eigenpy.hpp everywhere (this was discussed [in this PR on crocoddyl](https://github.com/loco-3d/crocoddyl/pull/1112))

The point about removing `#include <boost/python.hpp>` extends to including Boost.Python but not eigenpy.
Starting from [eigenpy 2.9.0](https://github.com/stack-of-tasks/eigenpy/commit/e9ba801d6563ce7a7aecaf9ca37eea54f5073714), a new header covering alignment issues to support full vectorization is included in `<eigenpy/eigenpy.hpp>` and *needs* to be included in order to get vectorization support everywhere.  
As a rule of thumb, Boost.Python should never be included directly.

ManifoldFR committed:
Fixes for more recent version of Boost + Pinocchio3x

+ using use new containers from new eigenpy 2.8.0

ManifoldFR committed:
remove ContactType enum, use crocoddyl::ContactType (from master branch)

ManifoldFR committed:
pre-commit: run all

ManifoldFR committed:
python/main.cpp: do not import pinocchio twice

ManifoldFR committed:
python/walk-with-traj: fix how std::vector is exposed

* exposed in the wrong place
* exposition code was exposing for the second time

ManifoldFR committed:
python/main.cpp: use namespace aliases

* there should be no conflict in this file
* use properly namespace'd eigenpy::enableEigenPySpecific function instead of macro

ManifoldFR committed:
fix warnings about fwd declaring a struct as "class"

ManifoldFR committed:
python/std-containers: fix pinocchio<2.9.2, eigenpy<2.8

ManifoldFR commented:
I think I should update the CI to also test a build with eigenpy 2.8.0

nim65s commented on .pre-commit-config.yaml:
```diff
@@ -5,7 +5,7 @@ repos:
     hooks:
     -   id: clang-format
-        args: [--style=Google]
+        args: ['--style={BasedOnStyle: Google, SortIncludes: false}']
```
I think sorting includes is a good practice. I understand that in older big projects, brutally moving all of them at once can cause tons of complications, but in a fairly new project where they are already sorted, I don't see the point of stopping this sort.

What is your motivation for this ?

ManifoldFR replied:
I changed this reflexively -- we did the same change in the PR stack-of-tasks/eigenpy#325 (at [this commit](stack-of-tasks/eigenpy@b482f27f))

What happened is that the pre-commit CI bot would sometimes do something like this and break things:
```cpp
#include <eigenpy/eigen-from-python.hpp>
#include <eigenpy/eigenpy.hpp>
```
unless we added a line with a comment
```cpp
#include <eigenpy/eigenpy.hpp>
// include after
#include <eigenpy/eigen-from-python.hpp>
```

nim65s replied:
In a perfect world, changing the include order wouldn't change the behavior. When it does, I think it is super important to document it, both for 
- the next human changing the code to warn they not to do a mistake here, 
- and for automated tools trying blindly "improving" the code

So I think we should keep this comment in this case.

Another (better but harder) solution would be to include only `<eigenpy/eigen-from-python.hpp>`, and let `<eigenpy/eigen-from-python.hpp>` include the `<eigenpy/eigenpy.hpp>`, or find another similar idea which would avoid breaking things if someone include those files in this order. Or at least detect that in eigenpy and warn the developer as soon and clearly as possible.

ManifoldFR replied:
Understood, I'll reverse that change then

nim65s reviewed (commented):
Thanks for this !
I'll foward this new eigenpy release soon.

nim65s commented:
Wow, this somehow broke the numerical behavior of a solver on 20.04, which now oscillates between 2 solutions forever: 
```
35941  2.08096e-01  3.30847e-03  3.59524e-13  6.93741e+07  6.93741e+07  1.0000  0.00000e+00
35942  2.06187e-01  3.30847e-03  3.59524e-12  6.93741e+08  6.93741e+08  0.0020  0.00000e+00
35943  2.08096e-01  3.30847e-03  3.59524e-13  6.93741e+07  6.93741e+07  1.0000  0.00000e+00
35944  2.06187e-01  3.30847e-03  3.59524e-12  6.93741e+08  6.93741e+08  0.0020  0.00000e+00
```

I'll re-run the CI on the current devel to be sure this comes from here:
https://gitlab.laas.fr/memory-of-motion/sobec/-/jobs/178651

ManifoldFR commented:
> Wow, this somehow broke the numerical behavior of a solver on 20.04, which now oscillates between 2 solutions forever:
> 
> ```
> 35941  2.08096e-01  3.30847e-03  3.59524e-13  6.93741e+07  6.93741e+07  1.0000  0.00000e+00
> 35942  2.06187e-01  3.30847e-03  3.59524e-12  6.93741e+08  6.93741e+08  0.0020  0.00000e+00
> 35943  2.08096e-01  3.30847e-03  3.59524e-13  6.93741e+07  6.93741e+07  1.0000  0.00000e+00
> 35944  2.06187e-01  3.30847e-03  3.59524e-12  6.93741e+08  6.93741e+08  0.0020  0.00000e+00
> ```
> 
> I'll re-run the CI on the current devel to be sure this comes from here: https://gitlab.laas.fr/memory-of-motion/sobec/-/jobs/178651

Seems like it passed according to your link. What happened ?

ManifoldFR commented:
Just did a *big* rebase on `devel`.

ManifoldFR committed:
crocomplements: fix registering shared_ptr of ResidualDataFlyAngle

std::vector of Vector3 shouldn't be exposed twice

ManifoldFR committed:
cmake : add GENERATE_PYTHON_STUBS option

ManifoldFR committed:
change exposed name of StdVectorStdStringIndex_

+ under StdVec_StdString (name adopted in future version of pinocchio)
+ yaml_params.py : do not try to use the exposed type in the if/else

Some future version of Pinocchio exposes std::vector<std::string>. This
change makes the code future-proof.

ManifoldFR committed:
python: remove useless #include <boost/python.hpp>

+ eigenpy's main header already includes the main boost/python header
+ including boost/enum is never necessary after boost/python

derp

ManifoldFR committed:
python/activation-quad-ref.cpp : remove useless include

+ crocoddyl/core/activation-base is already included by activation-quad-ref header

python/walk-with-traj: fix ocp.cpp including wrong header

ManifoldFR committed:
crocomplements/lowpassfilter: fix quasiStatic shadowing parent class virtual function

+ fix signature

ManifoldFR committed:
python : import all symbols from pywrap binary

ManifoldFR committed:
python : fix including vector_indexing_suite *after* eigenpy header

nim65s commented:
I think this can be merged, right ?

ManifoldFR commented:
I think so ! 😀

On Tue, 24 Jan 2023, 20:51 Guilhem Saurel, ***@***.***> wrote:

> I think this can be merged, right ?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/MeMory-of-MOtion/sobec/pull/72#issuecomment-1402509880>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AFA427E7Q3KINIAI7D3YIUDWUAXDZANCNFSM6AAAAAASVQHVTE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
